site/src/pages/release-notes/1.31.0.mdx

```diff
@@ -0,0 +1,168 @@
+  - You can create a <type://TlsProvider> which specifies <type://TlsKeyPair> dynamically.
+    ```java
+    // return a TlsKeyPair for host patterns
+    TlsProvider tlsProvider = TlsProvider
+      .builder()
+      .keyPair("*.a.com", TlsKeyPair.of(...))
+      .keyPair("*.b.com", TlsKeyPair.of(...))
+      .build();
+
+    // return a TlsKeyPair dynamically
+    TlsKeyPair tlsKeyPair = hostname -> TlsKeyPair.of(...);
+    ```
+  - The <type://TlsProvider> can be used for both server and clients.
+    ```java
+    // client-side
+    ClientFactory
+      .builder()
+      .tlsProvider(tlsProvider);
+
+    // server-side
+    Server
+      .builder()
+      .tlsProvider(tlsProvider);
+    ```
+- You can customize how in-flight requests are terminated during server shutdown by specifying a <type://GracefulShutdown>. #5941
+  ```java
+  GracefulShutdown gracefulShutdown =
+    GracefulShutdown.builder()
+                    .shutdownErrorFunction((ctx, req) -> {
+                        logger.info("Not gracefully terminated!");
+                        return new MyException();
+                    })
+                    .build();
+
+  Server
+    .builder()
+    .gracefulShutdown(gracefulShutdown);
+  ```
+- You can access the raw request path via <type://ServiceRequestContext#rawPath()>. #5931 #5932
+  ```java
+  ServiceRequestContext ctx = ...;
+  String rawPath = ctx.rawPath();
+    ```
+- More static factory methods are added for <type://ResponseEntity>. #5954
+  ```java
+  ResponseEntity response = ResponseEntity.of(200);
+  ResponseEntity response = ResponseEntity.of("Hello!");
+  ```
+- You can specify <type://HttpRequest#timeout(Duration)> or <type://HttpResponse#timeout(Duration)> to set a timeout
+  for an <type://HttpRequest> or <type://HttpResponse>. #5744 #5761
+  ```java
+  HttpRequest
+    .streaming(HttpMethod.GET, "/")
+    .timeout(Duration.ofSeconds(1)) // timeout if there is a delay exceeding 1 second between each data chunk
+  ```
+- You can now register your server to or retrieve <typeplural://Endpoint> from
+a [Nacos](https://nacos.io/) registry server. #5365 #5409
+  - Use <type://NacosUpdatingListener> to register the <typeplural://Server> to [Nacos](https://nacos.io/):
+  ```java
+  Server server =
+    Server.builder()
+          ...;
+  server.addListener(
+    NacosUpdatingListener.builder(nacosUri, "my-service")
+      ...
+      .build());
+    ```
+  - Use <type://NacosEndpointGroup> for clients:
+  ```java
+  EndpointGroup group = NacosEndpointGroup.of(nacosUri, "my-service");
+  WebClient client = WebClient.of(SessionProtocol.HTTP, group);
+  ```
+- An experimental `micrometer-context` module has been added. You can propagate <type://RequestContext> via [Project Reactor](https://projectreactor.io/)'s `Context`. #5145 #5577
+  ```java
+  ContextRegistry
+    .getInstance()
+    .registerThreadLocalAccessor(new RequestContextThreadLocalAccessor());
+  Hooks.enableAutomaticContextPropagation();
+  ```
```

Comment:
Could we remove this from the release notes since it is excluded from the publish?